src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -129,7 +168,54 @@ public static ODataItemBase ReadResourceOrResourceSet(this ODataReader reader)
                         }
 
                         break;
+                    case ODataReaderState.DeltaResourceSetStart:
+                        ODataDeltaResourceSet deltaResourceSet = (ODataDeltaResourceSet)reader.Item;
+                        Contract.Assert(deltaResourceSet != null, "ResourceSet should never be null.");
+
+                        ODataDeltaResourceSetWrapper deltaResourceSetWrapper = new ODataDeltaResourceSetWrapper(deltaResourceSet);
+                        if (itemsStack.Count > 0)
+                        {
+                            ODataNestedResourceInfoWrapper parentNestedResourceInfo = (ODataNestedResourceInfoWrapper)itemsStack.Peek();
+                            Contract.Assert(parentNestedResourceInfo != null, "this has to be an inner resource set. inner resource sets always have a nested resource info.");
+                            Contract.Assert(parentNestedResourceInfo.NestedResourceInfo.IsCollection == true, "Only collection nested properties can contain resource set as their child.");
+                            parentNestedResourceInfo.NestedItems.Add(deltaResourceSetWrapper);
+                        }
+                        else
+                        {
+                            topLevelItem = deltaResourceSetWrapper;
+                        }
+
+                        itemsStack.Push(deltaResourceSetWrapper);
+                        break;
 
+                    case ODataReaderState.DeltaResourceSetEnd:
+                        Contract.Assert(itemsStack.Count > 0 && itemsStack.Peek().Item == reader.Item, "The resource set which is ending should be on the top of the items stack.");
+                        itemsStack.Pop();
+                        break;
```

Comment:
Exact same code as ODataReaderState.ResourcesetEnd.  Just handle them together:

case ODataReaderState.ResourceSetEnd:
case ODataReaderState.DeltaResourceSetEnd:
        Contract.Assert(itemsStack.Count > 0 && itemsStack.Peek().Item == reader.Item, "The resource set which is ending should be on the top of the items stack.");
        itemsStack.Pop();
        break;